From 974d92cb1fcad6941e011602506a9a10055e15da Mon Sep 17 00:00:00 2001
From: Pau Ruiz Safont <pau.safont@vates.tech>
Date: Fri, 19 Dec 2025 12:17:23 +0000
Subject: [PATCH] xapi_vm_migrate: add capabilities to migration not supported
 by sr error

The error does not make it clear at all what are the capabilities that
are stopping it from supporting migration, which makes debugging issues
difficult, add a string that contains them. This can be easily ignored
by default in clients, but the information is easily available if needed.

Signed-off-by: Pau Ruiz Safont <pau.safont@vates.tech>
---
 ocaml/idl/datamodel_errors.ml | 2 +-
 ocaml/xapi/xapi_vm_migrate.ml | 8 ++++++--
 2 files changed, 7 insertions(+), 3 deletions(-)

diff --git a/ocaml/idl/datamodel_errors.ml b/ocaml/idl/datamodel_errors.ml
index 361a33270..c1b1f09e2 100644
--- a/ocaml/idl/datamodel_errors.ml
+++ b/ocaml/idl/datamodel_errors.ml
@@ -1059,7 +1059,7 @@ let _ =
     ~doc:"The VDI mirroring cannot be performed" () ;
   error Api_errors.too_many_storage_migrates ["number"]
     ~doc:"You reached the maximal number of concurrently migrating VMs." () ;
-  error Api_errors.sr_does_not_support_migration ["sr"]
+  error Api_errors.sr_does_not_support_migration ["sr"; "capabilities"]
     ~doc:"Cannot migrate a VDI to or from an SR that doesn't support migration."
     () ;
   error Api_errors.vm_failed_shutdown_ack ["vm"]
diff --git a/ocaml/xapi/xapi_vm_migrate.ml b/ocaml/xapi/xapi_vm_migrate.ml
index 374201466..5883eb821 100644
--- a/ocaml/xapi/xapi_vm_migrate.ml
+++ b/ocaml/xapi/xapi_vm_migrate.ml
@@ -162,8 +162,12 @@ open Storage_interface
 
 let supported_on ~ops sr sr_features =
   let open Smint.Feature in
-  if not (List.for_all (fun op -> has_capability op sr_features) ops) then
-    let msg = [Ref.string_of sr] in
+  let missing =
+    List.filter (fun op -> not (has_capability op sr_features)) ops
+  in
+  if missing <> [] then
+    let missing = List.map capability_to_string missing |> String.concat ";" in
+    let msg = [Ref.string_of sr; missing] in
     raise Api_errors.(Server_error (sr_does_not_support_migration, msg))
 
 let assert_sr_support_operations ~__context ~vdi_map ~remote ~local_ops
